### PR TITLE
Feature/unst 8490 621 extforce converter boundary condition

### PR DESCRIPTION
**Pull request by MAfarrag**
# Add Converter for the Boundary Condition quantities in the old external forcing file

## New Features
Now there is a boundary condition converter

## Design
- All the existing converters "initial condition, boundary condition, mete data" now exist in one file `hydrolib/tools/ext_old_to_new/converters.py` as there are a lot of similarities and most probably in the future they will be redesigned, or lots of the repeated lines will be moved upstream to the `BaseConverter` super class.
- The `ext_old_to_new` is redesigned and  replaced with the `ExternalForcingConverter` class
```
ExternalForcingConverter
|-__init__
|- read_old_file
|- update
|- save
|- extold_model
|- ext_model (setter and getter)
|- inifield_model (setter and getter)
|- structure_model (setter and getter)
``` 
- The backup parameter in the `hydrolib/tools/ext_old_to_new/utils.backup_file` funcion is remove as there is now need to enter the function if you actually do not want to backup you file (the function should be called only if you need its functionality) and the condition should be made in the calling script.

## Formating and Styling
- I added lots of formatters that helps to format your code locally before pushing to the `sonar` online tool, these formatter works standalone or with pre-commit hooks.
- The pre-commit hooks configuration file is not part of the repo for now.

### Data
- `tests/data/input/boundary-conditions/old-external-boundary_condition_only.ext`
- `tests/data/input/dflowfm_individual_files/polylines/polyline-no-z-no-label.pli`
- `tests/data/input/dflowfm_individual_files/polylines/polyline-no-z-with-label.pli`
- `tests/data/input/dflowfm_individual_files/polylines/polyline-with-z-no-label.pli``
- `tests/data/input/dflowfm_individual_files/polylines/polyline-with-z-no-label.pliz`

### New Tests
- Creating subfolders in the testing directory to replicate the same subfolders/modules of the package. `tests/dflowfm/extold/test_boundary.py` .
- Create `tests/dflowfm/extold/test_boundary.py::TestBoundary::test_existing_file`.
- Test the polyline with different cases `tests/dflowfm/polyfile/test_polyline_models.py`, currently the polyline failes with the scientific notation, and reads only the the number of dimensions (2*2) will read only the first two numbers in the row even if there is a z, and the z value will only be read if the second dimension is 3 and the extension has to  be `.pliz` in the same time.
- create a test for the converter `tests/tools/test_converters.py::TestBoundaryConverter`.
- create tests for all the methods and properties in the `ExternalForcingConverter` class.

## Old errors
- Multiple uses of mutable objects like lists  `[]` as a default values was converted to `Field(default_factory=list)`

### CODE REVIEW

**Review by arthurvd — CHANGES_REQUESTED** [review]
Hi, see my various comments. Maybe we need to talk tomorrow about the (silly) difference between regular polylines and boundary polylines.

File: hydrolib/tools/ext_old_to_new/converters.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            forcing.method
+        )
+        if block_data["interpolationmethod"] == InterpolationMethod.averaging:
+            block_data["averagingtype"] = oldmethod_to_averaging_type(forcing.method)
+            block_data["averagingrelsize"] = forcing.relativesearchcellsize
+            block_data["averagingnummin"] = forcing.nummin
+            block_data["averagingpercentile"] = forcing.percentileminmax
+        block_data["operand"] = forcing.operand
+
+        if hasattr(forcing, "extrapolation"):
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/tools/ext_old_to_new/converters.py:178]: converted the value as following 
```
if hasattr(forcing, "extrapolation"):
       block_data["extrapolationmethod"] = "yes" if forcing.extrapolation == 1 else "no"
```

File: hydrolib/tools/ext_old_to_new/main_converter.py [hunk 2]
```diff
@@ -27,109 +27,208 @@
...
-            The updated models (already written to disk). Maybe used
-            at call site to inspect the updated models.
-    """
+    @property
+    def extold_model(self):
+        """old external forcing model."""
+        return self._extold_model
+
+    @property
+    def ext_model(self) -> FileModel:
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/tools/ext_old_to_new/main_converter.py:87]: the type check always refers to the superclass not the subclass, so if i make it ExtModel the syntax checker will refer to it as a wrong return type

File: hydrolib/tools/ext_old_to_new/main_converter.py [hunk 3]
```diff
@@ -27,109 +27,208 @@
...
+            StructureModel, path
+        )
 
-    try:
-        extold_model = _read_ext_old_data(extoldfile)
-    except Exception as error:
-        print("The old external forcing file could not be read:", error)
-        return
+    @classmethod
+    def read_old_file(cls, extoldfile: PathOrStr) -> "ExternalForcingConverter":
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/tools/ext_old_to_new/main_converter.py:133]: Redesigned:
the `ExternalForcingConverter` class can be instantiated by `ExtOldModel` or a path to external forcing file and the `_read_old_file` now is used inside the constructor to read the external forcing file if path is given to the converter

File: pyproject.toml [hunk 4]
```diff
@@ -27,32 +27,48 @@ strenum = "^0"
...
+markupsafe = "<3.0.2"
 jupyter = "^1.0.0"
 ipykernel = "^6.15.0"
 # matplotlib >=3.8 supports numpy 2.0
 matplotlib = "^3.8"
 # xarray >=2024.6.0 supports numpy 2.0, but consider looser version if adding it as actual package dependencies
 xarray = ">=2024.6.0"
-flake8-pyproject = "*"
+
+#[tool.poetry.group.docs.dependencies]
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply pyproject.toml]: updated

File: pyproject.toml [hunk 5]
```diff
@@ -27,32 +27,48 @@ strenum = "^0"
...
+flake8 = "^7.1.1"
+flake8-bandit = "^4.1.1"
+flake8-bugbear = "^24.10.31"
+flake8-docstrings = "^1.6.0"
+flake8-rst-docstrings = "^0.3.0"
+flake8-pyproject = "^1.2.3"
+isort = "^5.8"
+pep8-naming = "*"
+darglint = "^1.8.1"
+reorder-python-imports = "^3.8.2"
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply pyproject.toml]: removed

File: tests/data/input/dflowfm_individual_files/polylines/polyline-no-z-with-label.pli
[the remark this answers is not in the document]
Comment (MAfarrag) [reply tests/data/input/dflowfm_individual_files/polylines/polyline-no-z-with-label.pli:1]: renamed

**Review by MAfarrag — COMMENTED** [review]
.

File: hydrolib/tools/ext_old_to_new/converters.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    def __init__(self):
+        super().__init__()
+
+    def convert(self, forcing: ExtOldForcing) -> InitialField:
+        """Convert an old external forcing block with meteo data to a Meteo
+        forcing block suitable for inclusion in a new external forcings file.
+
+        This function takes a forcing block from an old external forcings
+        file, represented by an instance of ExtOldForcing, and converts it
+        into a Meteo object. The Meteo object is suitable for use in new
```
[the remark this answers is not in the document]
Comment (arthurvd) [reply hydrolib/tools/ext_old_to_new/converters.py]: @MAfarrag: this point is still open/unresolved.

File: hydrolib/tools/ext_old_to_new/main_converter.py [hunk 7]
```diff
@@ -27,109 +27,208 @@
...
-    return extold_model, ext_model, inifield_model, structure_model
+        if _verbose:
+            print(f"Read {len(extold_model.forcing)} forcing blocks from {extoldfile}.")
+
+        return cls(extold_model)
+
+    def update(
+        self,
+        postfix: str = "",
+    ) -> Union[Tuple[FileModel, FileModel, FileModel], None]:
```
[the remark this answers is not in the document]
Comment (MAfarrag) [reply hydrolib/tools/ext_old_to_new/main_converter.py]: this is now fixed

**Review by arthurvd — APPROVED** [review]
